### DISCUSSION TITLE
vcpkg compliant CMake example project for shared/static cross-platform builds?

### DISCUSSION THREAD

**Discussion by SamuelMarks** [microsoft/vcpkg#23093] [General]
Given (essentially) this directory layout:
```
|-- CMakeLists.txt
|-- libcurl-simple-https
|   |-- CMakeLists.txt
|   |-- curl_simple_https.c
|   `-- curl_simple_https.h
`-- vcpkg.json
```

For reuse as boilerplate across my decent number of C projects, I took https://gitlab.kitware.com/cmake/cmake/-/blob/master/Help/guide/tutorial/Complete :

  - replaced literal names with variables;
  - removed functionality irrelevant to my purposes;
  - added stronger `use strict` style guards:
    - [`set(CMAKE_C_STANDARD 90)`](https://cmake.org/cmake/help/latest/variable/CMAKE_C_STANDARD.html)
    - [`set(CMAKE_C_VISIBILITY_PRESET hidden)`](https://cmake.org/cmake/help/latest/prop_tgt/LANG_VISIBILITY_PRESET.html)
    - [`set(CMAKE_VISIBILITY_INLINES_HIDDEN YES)`](https://cmake.org/cmake/help/latest/variable/CMAKE_VISIBILITY_INLINES_HIDDEN.html)
    - `"$<${gcc_like}:$<BUILD_INTERFACE:-Wall;-Wextra;-Wshadow;-Wformat=2;-Wunused;-pedantic>>"`
    - `"$<${msvc}:$<BUILD_INTERFACE:-W3;-WX;-Zi;-permissive->>"`
  - generated export headers with [`GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html);
  - [`set`](https://cmake.org/cmake/help/latest/command/set.html) separate `Source_Files` and `Header_Files` vars [with [`source_groups`](https://cmake.org/cmake/help/latest/command/source_group.html)] for installation [and IDE assist]
  - added the [`INSTALL(EXPORT`](https://cmake.org/cmake/help/latest/command/install.html#export) lines.

Unfortunately where I've got up to so far isn't far enough…

### [`CMakeLists.txt`](https://github.com/SamuelMarks/curl-simple-https/blob/master/CMakeLists.txt)
```cmake
# control where the static and shared libraries are built so that on windows
# we don't need to tinker with the path to run the executable
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")

if (NOT DEFINED CMAKE_BUILD_TYPE)
    if (DEFINED CMAKE_RC_FLAGS_RELEASE)
        set(CMAKE_BUILD_TYPE "Release")
    else ()
        set(CMAKE_BUILD_TYPE "Debug")
    endif ()
endif (NOT DEFINED CMAKE_BUILD_TYPE)

if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set(INSTALL_LIB_DESTINATION "bin")
    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
        set(INSTALL_LIB_DESTINATION "${CMAKE_BUILD_TYPE}/${INSTALL_LIB_DESTINATION}")
    endif (CMAKE_BUILD_TYPE STREQUAL "Debug")
else()
    set(INSTALL_LIB_DESTINATION "lib")
endif (CMAKE_SYSTEM_NAME STREQUAL "Windows")
```

### [`libcurl-simple-https/CMakeLists.txt`](https://github.com/SamuelMarks/curl-simple-https/blob/master/libcurl-simple-https/CMakeLists.txt):
```cmake
set(installable_libs
        "${LIBRARY_NAME}" "${PROJECT_NAME}_compiler_flags" "${PROJECT_NAME}_utils" "${PROJECT_NAME}_google_auth")
if (TARGET "${DEPENDANT_LIBRARY}")
  list(APPEND installable_libs "${DEPENDANT_LIBRARY}")
endif (TARGET "${DEPENDANT_LIBRARY}")
install(TARGETS ${installable_libs}
        DESTINATION "${INSTALL_LIB_DESTINATION}"
        EXPORT "${LIBRARY_NAME}Targets")
install(FILES "${_export_file}" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}_configure.h" ${Header_Files}
        DESTINATION "include")
install(EXPORT "${LIBRARY_NAME}Targets" DESTINATION "share/${LIBRARY_NAME}")
```

All my attempts at creating simple C projects that were subsequently offered as vcpkg ports:
 - https://github.com/SamuelMarks/curl-simple-https
 - https://github.com/offscale/cauthflow (see its `CMakeLists.txt` I tried setting `RUNTIME`/`LIBRARY`/`ARCHIVE` `DESTINATION` and different global `OUTPUT_DIRECTORY`)
 - https://github.com/offscale/c89randomstr (works but is header-only)
 - https://github.com/offscale/c89stringutils (works but is header-only)

What am I doing wrong? - Is there an example project boilerplate I can follow? - Even the oft-copied `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")` in my `portfile.cmake` seems wrong…

Thanks for suggestions

**Comment by Neumann-A** [comment]
- Use https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html
- Use `install(FILES)` with the `TYPE` signature
- Use install(EXPORT) to generate  targets.cmake
- use https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html

Don't do
- target based compiler interface helpers: `"${PROJECT_NAME}_compiler_flags"`
- don't set `CMAKE_BUILD_TYPE`. leave this up to the user. 
- no need for config aware `INSTALL_LIB_DESTINATION`. (vcpkg automatically runs for both buildtypes and adjust the prefix so you are not helping here)

Notes:
- generated export headers with [GenerateExportHeader](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html); Might require a `vcpkg_string_replace` to hardcode static vs shared builds. (Mostly static because shared is the common default )

> file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")

This is required by header only ports. Alternatively you can set `VCPKG_BUILD_TYPE` in the portfile to `release` but you need to make sure that your exported targets are aware of that and configuration independent. .

**Reply by SamuelMarks** [reply]
Ok thanks. I think I've got the hang of that… maybe, just updated my PR: https://github.com/microsoft/vcpkg/pull/22917

**Reply by SamuelMarks** [reply]
@Neumann-A No, I am still getting:
```
The following dlls were found in /lib or /debug/lib. Please move them to /bin or /debug/bin, respectively.

    D:/packages/libcurl-simple-https_x86-windows/lib/libcurl-simple-https.dll

Mismatching number of debug and release binaries. Found 0 for debug but 1 for release.
Debug binaries


Release binaries

    D:/packages/libcurl-simple-https_x86-windows/lib/libcurl-simple-https.lib

Debug binaries were not found
```

**Reply by Neumann-A** [reply]
Your 
```
install(TARGETS ${installable_libs}
        DESTINATION "${INSTALL_LIB_DESTINATION}"
        EXPORT "${LIBRARY_NAME}Targets")
```
is not enough to install targets/libraries/executables correctly. 
You can try without `DESTINATION`
```
install(TARGETS ${installable_libs} EXPORT "${LIBRARY_NAME}Targets")
```
and hope cmake does the right thing or use 
`ARCHIVE|LIBRARY|RUNTIME`

```
install(TARGETS ${installable_libs} EXPORT "${LIBRARY_NAME}Targets"
          ARCHIVE DESTINATION <libdir>
          LIBRARY DESTINATION <libdir>
          RUNTIME DESTINATION <bindir>
          )
```

Please watch https://www.youtube.com/watch?v=m0DwB4OvDXk (the install stuff is around 30 min)

**Reply by SamuelMarks** [reply]
Updated both ports [libcurl-simple-https, cauthflow] with open PRs. Didn't fix the bug. Will watch that YouTube video later tonight.

**Reply by Neumann-A** [reply]
`cmake_minimum_required(VERSION 3.0)` .... but that is not the issue here. 

It cannot work if you still delete the `/debug` folder in the portfile.cmake .....dlls are now in /bin and not in /lib.....
